src/codelabs/null-safety.md

```diff
@@ -0,0 +1,321 @@
+---
+title: Null safety codelab
+description: Learn about and practice writing null safe code in DartPad!
+js: [{url: 'https://dartpad.dev/inject_embed.dart.js', defer: true}]
+---
+{% assign useIframe = false -%}
+<?code-excerpt replace="/ *\/\/\s+ignore_for_file:[^\n]+\n//g; /(^|\n)*\/\/\s+ignore:[^\n]+\n/$1/g; /(\n[^\n]+) *\/\/\s+ignore:[^\n]+\n/$1\n/g"?>
+<?code-excerpt plaster="none"?>
+
+This codelab teaches you about Dart’s null-safe type system,
+which was introduced in Dart SDK 2.12. When you opt into null safety,
+types in your code are non-nullable by default,
+meaning that values can’t be `null` unless you say they can be.
+
+This codelab covers the following material:
+
+* Nullable and non-nullable types.
+* How and when to use the null-aware operators: `?` and `!`
+* How the `late` keyword affects variables and initialization.
+
+Using embedded DartPad editors, you can test your knowledge by
+running example code and completing exercises. To get the most out of
+this codelab, you should have some knowledge of [basic Dart syntax](/samples).
+
+{{site.alert.note}}
+  This page uses embedded DartPads to display examples and exercises.
+  {% include dartpads-embedded-troubleshooting.md %}
+{{site.alert.end}}
+
+## What is null safety?
+
+Dart's null-safe type system makes types in your code non-nullable by default.
+That means values in your Dart programs can't be `null` unless you say they can be.
+With null safety, you can find null reference errors before you even
+run your program.
+
+Here are some core principles of sound null safety in the Dart language:
```

Comment:
These's aren't really the "core principles" of null safety; we have those here:
https://dart.dev/null-safety#null-safety-principles

Can we just say something a bit more general like "The core traits of sound null safety in Dart are:" ?